src/main/scala/com/fulcrumgenomics/umi/DuplexConsensusCaller.scala

```diff
@@ -0,0 +1,255 @@
+    */
+  override protected def consensusSamRecordsFromSamRecords(recs: Seq[SAMRecord]): Seq[SAMRecord] = {
+    // Group the reads by /A vs. /B and then sort so that the larger group is first, and then by suffix
+    val groups = recs.filter(_.getReadPairedFlag)
+      .groupBy(_.getStringAttribute(ConsensusTags.MolecularId))
+      .toSeq.sortWith {
+      case ((lhsMi, lhsRecs), (rhsMi, rhsRecs)) =>
+        if (lhsRecs.size > rhsRecs.size) true
+        else if (rhsRecs.size > lhsRecs.size) false
+        else lhsMi < rhsMi
+    }.map(_._2)
+
+    groups match {
+      case Seq() =>
+        Nil
+      case Seq(ab) =>
+        rejectRecords(ab)
+        Nil
+      case Seq(ab, ba) =>
+        // Fragments have no place in duplex land!
+        val (_, abR1s, abR2s) = subGroupRecords(ab)
+        val (_, baR1s, baR2s) = subGroupRecords(ba)
+
+        // Filter by common indel pattern with AB and BA together
+        val filteredXs = filterToMostCommonAlignment(abR1s ++ baR2s)
+        val filteredYs = filterToMostCommonAlignment(abR2s ++ baR1s)
+
+        // Then split then back apart for SS calling
+        val filteredAbR1s = filteredXs.filter(_.getFirstOfPairFlag)
+        val filteredAbR2s = filteredYs.filter(_.getSecondOfPairFlag)
+        val filteredBaR1s = filteredYs.filter(_.getFirstOfPairFlag)
+        val filteredBaR2s = filteredXs.filter(_.getSecondOfPairFlag)
+
+        // Call the single-stranded consensus reads
+        val abR1Consensus = ssCaller.consensusFromSamRecords(filteredAbR1s)
+        val abR2Consensus = ssCaller.consensusFromSamRecords(filteredAbR2s)
+        val baR1Consensus = ssCaller.consensusFromSamRecords(filteredBaR1s)
+        val baR2Consensus = ssCaller.consensusFromSamRecords(filteredBaR2s)
+
+        // Call the duplex reads
+        val duplexR1 = duplexConsensus(abR1Consensus, baR2Consensus)
+        val duplexR2 = duplexConsensus(abR2Consensus, baR1Consensus)
+
+        // Convert to SAMRecords and return
+        (duplexR1, duplexR2) match {
+          case (Some(r1), Some(r2)) =>
+            Seq(createSamRecord(r1, FirstOfPair), createSamRecord(r2, SecondOfPair))
+          case _                    =>
+            rejectRecords(recs)
+            Nil
+        }
+      case gs =>
+        unreachable("SAMRecords supplied with more than two distinct MI values.")
+    }
+  }
+
+  /**
+    * Constructs a duplex consensus read from a pair of single strand consensus reads.
+    * If either of the incoming reads are undefined, the duplex read will be undefined.
+    *
+    * @param ab the single-strand consensus from one strand
+    * @param ba the single-strand consensus from the other strand
+    * @return a duplex consensus if one can be built
+    */
+  private[umi] def duplexConsensus(ab: Option[VanillaConsensusRead], ba: Option[VanillaConsensusRead]): Option[DuplexConsensusRead] = {
+    (ab, ba) match {
+      case (Some(a), Some(b)) =>
+        val len = min(a.length, b.length)
```

Comment:
Yeah.  It's complicated to describe though ;)  Basically what you get back is the longest length at which you had at least one AB R1 and one BA R2 (for read 1) and vice versa.  I feel like trying to explain that might cause more confusion than help.  I'll see what I can do though.